File: science-unlicense-format-tga/src/tgaformat.rs
//
// Public Domain - unlicense.science
//

use science_unlicense_encoding::api::store::{Format, Store};
use science_unlicense_encoding::api::io::Readable;
use std::error::Error;
use std::fs::File;
use crate::TGAReader;

pub struct TGAFormat {

}

impl TGAFormat {

    pub fn new() -> Self{
        return TGAFormat{};
    }
}

impl Format for TGAFormat {

    fn get_identifier(&self) -> &str {
        return "tga";
    }

    fn get_short_name(&self) -> &str {
        return "tga";
    }

    fn get_long_name(&self) -> &str {
        todo!()
    }

    fn get_mime_types(&self) -> &Vec<&str> {
        todo!()
    }

    fn get_extensions(&self) -> &Vec<&str> {
        todo!()
    }

    fn get_signature(&self) -> &Vec<&Vec<u8>> {
        todo!()
    }

    fn is_subset_format(&self) -> bool {
        todo!()
    }

    fn get_resource_types(&self) -> &Vec<&str> {
        todo!()
    }

    fn can_decode(&self, _input: &mut dyn Readable) -> Result<bool, Box<dyn Error>> {
        todo!()
    }

    fn search_end(&self, _input: &mut dyn Readable, _fullscan: bool) -> Result<u64, Box<dyn Error>> {
        todo!()
    }

    fn get_parameters(&self) {
        todo!()
    }

    fn open(&self, _source: File) -> Result<Box<dyn Store>, Box<dyn Error>> {
        let _reader = TGAReader::new();
        todo!()
    }
}

File: science-unlicense-format-tga/src/TGAFormat.java

package science.unlicense.format.tga;

import science.unlicense.common.api.Arrays;
import science.unlicense.common.api.character.Chars;
import science.unlicense.common.api.number.Endianness;
import science.unlicense.encoding.api.io.ByteInputStream;
import science.unlicense.encoding.api.io.DataInputStream;
import science.unlicense.encoding.api.io.IOException;
import science.unlicense.encoding.api.store.Store;
import science.unlicense.image.api.AbstractImageFormat;

/**
 *
 * @author Johann Sorel
 */
public class TGAFormat extends AbstractImageFormat {

    public static final TGAFormat INSTANCE = new TGAFormat();

    private TGAFormat() {
        super(new Chars("tga"));
        shortName = new Chars("TGA");
        longName = new Chars("Truevision Targa");
        mimeTypes.add(new Chars("image/tga"));
        mimeTypes.add(new Chars("image/targa"));
        mimeTypes.add(new Chars("application/targa"));
        extensions.add(new Chars("tga"));
        extensions.add(new Chars("tpic"));
    }

//    public boolean canDecode(Object input) throws IOException {
//        if (super.canDecode(input)){
//            return true;
//        } else {
//            final boolean[] closeStream = new boolean[1];
//            final ByteInputStream stream = IOUtilities.toInputStream(input, closeStream);
//            try{
//                return canDecode(stream);
//            }finally{
//                if (closeStream[0]){
//                    stream.close();
//                }
//            }
//        }
//    }

    private boolean canDecode(ByteInputStream stream) throws IOException {
        final DataInputStream ds = new DataInputStream(stream, Endianness.BIG_ENDIAN);
        final byte[] buffer = new byte[3];
        ds.readFully(buffer);
        //not exact but we don't want to read the file footer to know if it's a tga
        if (Arrays.contains(new int[]{
            TGAMetaModel.IMAGE_TYPE_NODATA,
            TGAMetaModel.IMAGE_TYPE_RLE_BW,
            TGAMetaModel.IMAGE_TYPE_RLE_COLORMAP,
            TGAMetaModel.IMAGE_TYPE_RLE_TRUECOLOR,
            TGAMetaModel.IMAGE_TYPE_UNCOMPRESSED_BW,
            TGAMetaModel.IMAGE_TYPE_UNCOMPRESSED_COLORMAP,
            TGAMetaModel.IMAGE_TYPE_UNCOMPRESSED_TRUECOLOR
            }, buffer[2])){
            //TODO we should do more checks
            return true;
        }
        return false;
    }

    @Override
    public boolean supportReading() {
        return true;
    }

    @Override
    public boolean supportWriting() {
        return false;
    }

    @Override
    public Store open(Object source) throws IOException {
        return new TGAStore(this, source);
    }

}
